### PR TITLE
Use mmw-geoprocessing

### PR DESCRIPTION
**Pull request by jamesmcclain**
The Geotrellis-based service that runs on top of spark-jobserver is now used to produce the "survey" of the AOI that appears on the `/analyze` page, and to produce the input for the TR-55 model for the `/project/x/scenario/y` pages.

Connects #569 
Connects #585 
Connects #701 
Supersedes #621 

**New Setup Instructions**

   * Make sure that your vagrant setup has an up-to-date base box by typing `vagrant box update`
   * Put the S3 credentials into your `~/.aws/credentials` file, those credentials will be copied into the worker VM when it is provisioned
      * Look-up S3 credentials on the fileshare
      * Install the aws cli tools if you have not done so already: `pip install awscli`
      * Create a profile called `mmw-stg` using the aws cli tools and the information from the fileshare (type `aws configure --profile mmw-stg`, more information [here](http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-quick-configuration)).
   * Destroy the `app` and `worker` VMs
   * Apply this change set
   * Rebuild the `app` and `worker` VMs

**To Test**
   1. Use the MMW site, in particular the `/analyze` and `/project` pages.

**Caveats**
   1. Requests occasionally time out.  This is seen more with TR-55 than on the `/analyze` page.  The more BMPs/reclassifications there are, the more likely there is to be a problem.  If you reload the page a few times, you should eventually get your results.
   2. Curiously, sometimes results (such as the bar charts) simply fail to appear with no error message in the console.  If you reload a few times, you should eventually get your results.
   3. ~~I moved the tests that I had for #589 and #604 into a separate branch.  I would like to keep the tests in a separate branch since they do not work on Jenkins, right now.~~
   4. Water quality calculations may not be correct yet.

aoi
screenshot from 2015-08-19 16 36 34
screenshot from 2015-08-19 16 37 49
screenshot from 2015-08-19 16 38 06
screenshot from 2015-08-19 16 38 28

---

### Streamlining Tasks

- [x] Mount virtual machine host AWS credentials into virtual machine when developing and testing
- [x] Cut `0.2.0` release of `tr55` Python module and use it
- [x] Create a container image that uses the Azavea fork of SJS with Joda-Time fix
- [x] Setup Docker on `worker` virtual machine
- [x] Assemble reproducible GeoTrellis JARs
- [x] Setup Upstart service definition for SJS container
- [x] Figure out how to build and reference data hub JAR
- [x] Automate uploading data hub JAR to SJS

---

**Old Setup Instructions**

The following instructions assume that you will be running the datahub service on your actual hardware, but they should work just as well if you run it in a VM or a container (or a container in a VM).

   1. Put the aws credentials for the S3 bucket into your environment or your ~/.aws/credentials file.
   2. This code requires version 0.2.2 of the TR-55 code.  You must either reprovision the `app` and `worker` VMs, or you must `ssh` into each of them and type `sudo pip uninstall tr55` (to uninstall the old code) and `sudo pip install git+https://github.com/WikiWatershed/tr-55.git` (to install the new code).
   3. Obtain, compile, and run the spark-jobserver.
      1. The version from https://github.com/jamesmcclain/spark-jobserver will allow you to run the datahub code without supplying additional configuration, but the version from https://github.com/azavea/spark-jobserver/releases/tag/v0.5.2-azavea can also be used.
      4. Compile the spark-jobserver: from within the jobserver src directory, type: `sbt 'project job-server' compile assembly`
      5. Run the spark-jobserver: from within the jobserver src directory type: `$SPARK_HOME/bin/spark-submit job-server/target/scala-2.10/spark-job-server.jar` (this assumes that you have spark installed, of course).
   4. Obtain the datahub code from here https://github.com/jamesmcclain/datahub-services/ and the Geotrellis code from here https://github.com/geotrellis/geotrellis.  The datahub code is known to work with Geotrellis SHA cb236ac.  After Geotrellis has been published locally, go into the datahub src directory and type `sbt assembly` to build the datahub code. 
   5. Upload the datahub jar to the spark-jobserver:
      1. Type `curl --data "" 'http://localhost:8090/contexts/summary-context'` to create a new spark context
      2. Type `curl --data-binary @summary/target/scala-2.10/datahub-summary-assembly-0.0.1.jar 'http://localhost:8090/jars/summary'` to upload the jar to the jobserver
      3. Type `curl --data-binary @examples/request.json 'http://localhost:8090/jobs?sync=true&context=summary-context&appName=summary&classPath=com.azavea.datahub.SummaryJob'` to upload the example request `example/request.json` to the jobserver.  If you see output that includes the string `"status": "OK"`, then things are probably working.
   6. Create an ssh tunnel between port 8090 of the host machine (where the datahub service is running) and port 8090 of the worker VM: `ssh -R 8090:localhost:8090 vagrant@33.33.34.20`

### CODE REVIEW

**Comment by jamesmcclain** [comment]
@azaveaci rebuild

File: src/mmw/apps/modeling/geoprocessing.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+from __future__ import absolute_import
+
+from django.conf import settings
+
+import httplib
+import json
+
+
+NLCD_MAPPING = {
+    11: ['water', 'Water'],
```
Comment (caseypt) [inline src/mmw/apps/modeling/geoprocessing.py]: Do these values line up with the values in the raster? Consider adding a comment.

File: src/mmw/apps/modeling/geoprocessing.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    82: ['row_crop', 'Row Crop'],
+    90: ['woody_wetland', 'Woody Wetland'],
+    95: ['herbaceous_wetland', 'Herbaceous Wetland']
+}
+
+SOIL_MAPPING = {
+    1: ['a', 'A'],
+    2: ['b', 'B'],
+    3: ['c', 'C'],
+    4: ['d', 'D']
```
Comment (caseypt) [inline src/mmw/apps/modeling/geoprocessing.py]: Here as well.

File: src/mmw/apps/modeling/geoprocessing.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+            return [[]]
+    except Exception:
+        conn.close()
+        return [[]]
+
+
+def geojson_to_x(data, nucleus, update_rule, after_rule):
+    """
+    Transform a raw Geotrellis histogram into an object of the desired
+    form (dictated by the last three arguments).
```
Comment (caseypt) [inline src/mmw/apps/modeling/geoprocessing.py]: Consider adding some example usages to the docstring.

File: src/mmw/apps/modeling/tasks.py [hunk 4]
```diff
@@ -28,110 +27,14 @@ def run_analyze(area_of_interest):
     average_lat = reduce(lambda total, p: total+p[1], pairs, 0) / len(pairs)
 
     # Zoom level is a hard coded value in the geoprocessing library.
-    zoom = 13
+    zoom = 11
+    tile_width = 768
```
Comment (caseypt) [inline src/mmw/apps/modeling/tasks.py]: Consider adding a comment describing the significance of this value.

File: src/mmw/apps/modeling/geoprocessing.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    host = settings.GEOP['host']
+    host = 'localhost'  # XXX
+    port = settings.GEOP['port']
+    path = settings.GEOP['path']
+    request = settings.GEOP['request'].copy()
+    request['input']['polyMask'] = polygon
+    conn = httplib.HTTPConnection("%s:%s" % (host, port))
+
+    try:
+        conn.request('POST', path, json.dumps(request))
```
Comment (caseypt) [inline src/mmw/apps/modeling/geoprocessing.py]: Consider moving the connection code into a helper function.

**Comment by caseypt** [comment]
I haven't had a chance to pull this down yet, but nice work. My main comment is that you should consider adding more comments to undocumented functions, and usage examples to the more complicated functions. 

Also, as @hectcastro suggested, I think we should push the new release of the TR-55 code and update this PR to use the new version before we merge this in. This will help reduce the amount of steps needed to test this and to reduce the amount of temporary code (i.e. sourcing the library from Github) and get it closer to it's final state.

**Comment by caseypt** [comment]
Also, what's the relationship between all of the dependencies and the `geoprocessing` VM? Should that VM be provisioned with all of these things?

**Comment by jamesmcclain** [comment]
> Also, what's the relationship between all of the dependencies and the geoprocessing VM? Should 
> that VM be provisioned with all of these things?

As it stands right now, the `geoprocessing` VM does not even need to be running for this to work (that is, if you are running the geotrellis stuff on your host machine -- you could run it in the VM if you wanted to).

**Comment by jamesmcclain** [comment]
> I haven't had a chance to pull this down yet, but nice work. My main comment is that you should 
> consider adding more comments to undocumented functions, and usage examples to the more 
> complicated functions.

Okay, will-do.

> Also, as @hectcastro suggested, I think we should push the new release of the TR-55 code and 
> update this PR to use the new version before we merge this in. This will help reduce the amount of 
> steps needed to test this and to reduce the amount of temporary code (i.e. sourcing the library 
> from Github) and get it closer to it's final state.

I expect that it will be a while before this is merged.

File: src/mmw/requirements/base.txt [hunk 6]
```diff
@@ -9,4 +9,4 @@ django-registration-redux==1.2
 python-omgeo==1.7.2
 rauth==0.7.1
 djangorestframework-gis==0.8.2
-tr55==0.1.0
\ No newline at end of file
+git+https://github.com/WikiWatershed/tr-55.git
```
Comment (hectcastro) [inline src/mmw/requirements/base.txt]: We should now be able to use version `0.2.0` here.

Comment (jamesmcclain) [inline src/mmw/requirements/base.txt]: Cool, will change.  Many thanks.

**Comment by kdeloach** [comment]
Nice code and concise doc comments. Consider adding simple tests for these histogram functions.

File: src/mmw/apps/modeling/geoprocessing.py (hunk 5 above)
Comment (jamesmcclain) [inline src/mmw/apps/modeling/geoprocessing.py]: Okay, will see if I can reorganize this a little bit.

**Comment by jamesmcclain** [comment]
A newer version of this can be found here: https://github.com/jamesmcclain/model-my-watershed/tree/feature/use-datahub-no-collections .  I have not merged that branch into this one because it uses an newer version of the datahub-services code: https://github.com/jamesmcclain/datahub-services/tree/feature/no-more-geometry-collections .

**Comment by hectcastro** [comment]
We still don't really have a release for the datahub project, so I think should you should be able to update this branch with the `feature/use-datahub-no-collections` branch.

**Comment by hectcastro** [comment]
@azavea-bot rebuild

**Comment by hectcastro** [comment]
@azavea-bot rebuild

**Comment by lliss** [comment]
I tested this out. Removed the geoprocessing VM and rebuilt the worker. Provisioning failed the first time but I reran and it succeeded. It does appear to be doing proper analysis of the selected area and is returning plausible numbers for the models. Looks good.

**Comment by caseypt** [comment]
Very cool. I got it working as well. Nice work @jamesmcclain and @hectcastro.

For the analyze page, I did some basic checking with the NLCD overlay on, and the results look good. Also, making changes to scenarios using the land cover tools returns accurate (or at least expected) results.

One issue I noticed, and I'm not exactly sure where the problem may be, but if I add a conservation practice to a scenario, the model fails to return any results (it polls until timeout). I tried different conservation practices, conservation practices alone and with land cover changes, and none of these combinations worked.

**Comment by hectcastro** [comment]
> Provisioning failed the first time but I reran and it succeeded.

What was the failure?

**Comment by jamesmcclain** [comment]
> One issue I noticed, and I'm not exactly sure where the problem may be, but if I add a conservation practice to a scenario, the model fails to return any results (it polls until timeout). I tried different conservation practices, conservation practices alone and with land cover changes, and none of these combinations worked.

There are two possible issues:
   * There are some BMPs which are incompatible with certain soil types.  One BMP which should always work no matter what is the "Green roof" BMP.
   * There could be another TR-55 bug (like the one fixed by https://github.com/WikiWatershed/tr-55/pull/40 ).  May I follow up with you a little bit later on this issue?

**Comment by mmcfarland** [comment]
@jamesmcclain Is the BMP::Soil issue one of validation, then?  If so, could you amend https://github.com/WikiWatershed/tr-55/issues/34 to enumerate the limitations?  We'll need to come up with a strategy for enforcing constraints.

**Comment by mmcfarland** [comment]
Provisioning failed the first time for me as well with:

```
TASK: [model-my-watershed.geoprocessing | Ensure geoprocessing service was started] ***
 failed: [worker] => {"failed": true}
 msg: start: Job failed to start
```

Going again through the second time around seems to be hanging on
`TASK: [model-my-watershed.spark-jobserver | Ensure Spark Job Server is running] ***`

Might need to try again for a third time.

**Comment by hectcastro** [comment]
There may be some useful information in `/var/log/upstart/spark-jobserver.log`. This step is responsible for pulling down the SJS container image if it doesn't exist locally. That download duration may be causing some kind of timeout.

**Comment by mmcfarland** [comment]
Third time did it, after a fresh `destroy` && `provision`.  The jobserver log was a bit hard to parse, but possibly some meaningful output:

```
Error response from daemon: Cannot kill container spark-jobserver: no such id: spark-jobserver
Error: failed to kill containers: [spark-jobserver]
Error response from daemon: no such id: spark-jobserver
Error: failed to remove containers: [spark-jobserver]
```

There are no timestamps, so this might just be a kill before the job server is launched successfully.

**Comment by hectcastro** [comment]
> ```
Error response from daemon: Cannot kill container spark-jobserver: no such id: spark-jobserver
Error: failed to kill containers: [spark-jobserver]
Error response from daemon: no such id: spark-jobserver
Error: failed to remove containers: [spark-jobserver]
```

Error messages like this are expected. They can happen in a scenario where there is no SJS instance, but the Upstart script tries to ensure an SJS is down before it tries to bring up another.

**Comment by mmcfarland** [comment]
This is largely working for me, with the caveats listed above.  I'm really interested to see how this performs on an EC2 instance when it's closer to the S3 bucket.  I'm in favor of merging this and then creating/leaning on some issues along the lines of:

* Verify that input changes don't result in recomputing the tile census (seems like maybe it is)
* Check the water quality numbers, the output doesn't match my expectations
* Add some logging/metrics to help figure out which components are failing (client time out, geop service time out, bad input etc) and get some client facing representation of the same.

Nice work getting this running, everyone.
